sw/airborne/modules/computer_vision/opticflow/visual_estimator.c

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright (C) 2014 Hann Woei Ho
+ *
+ * This file is part of Paparazzi.
+ *
+ * Paparazzi is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * Paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @file modules/computer_vision/opticflow/visual_estimator.c
+ * @brief optical-flow based hovering for Parrot AR.Drone 2.0
+ *
+ * Sensors from vertical camera and IMU of Parrot AR.Drone 2.0
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+
+// Own Header
+#include "visual_estimator.h"
+
+// Computer Vision
+#include "opticflow/optic_flow_int.h"
+#include "opticflow/fast9/fastRosten.h"
+
+// for FPS
+#include "modules/computer_vision/opticflow_module.h"
+
+// Paparazzi Data
+#include "state.h"
+#include "subsystems/abi.h"
+
+// Downlink
+#include "subsystems/datalink/downlink.h"
+
+// Timer
+#include <sys/time.h>
+
+// Image size set at init
+unsigned int imgWidth, imgHeight;
+
+// Local variables
+unsigned char *prev_frame, *gray_frame, *prev_gray_frame;
+int old_img_init;
+float OFx, OFy, dx_sum, dy_sum;
+
+// ARDrone Vertical Camera Parameters
+#define FOV_H 0.67020643276
+#define FOV_W 0.89360857702
+#define Fx_ARdrone 343.1211
+#define Fy_ARdrone 348.5053
+
+// Corner Detection
+int *x, *y;
+int count = 0;
+int max_count = 25;
+#define MAX_COUNT 100
+
+// Corner Tracking
+int *new_x, *new_y, *status, *dx, *dy;
+int error_opticflow;
+int flow_count = 0;
+int remove_point;
+int c;
+int borderx = 24, bordery = 24;
+
+// Remove bad corners
+float distance2, min_distance, min_distance2;
+
+// Flow Derotation
+#define FLOW_DEROTATION
+float curr_pitch, curr_roll, curr_yaw, prev_pitch, prev_roll;
+float cam_h, diff_roll, diff_pitch, OFx_trans, OFy_trans;
+
+// Lateral Velocity Computation
+float Velx, Vely;
+
+struct FloatVect3 V_body;
+
+/** height above ground level, from ABI
+ * Used for scale computation, negative value means invalid.
+ */
+volatile float estimator_agl;
+
+/** default sonar/agl to use in opticflow visual_estimator */
+#ifndef OPTICFLOW_AGL_ID
+#define OPTICFLOW_AGL_ID ABI_BROADCAST
+#endif
+abi_event agl_ev;
+static void agl_cb(uint8_t sender_id, const float *distance);
+
+static void agl_cb(uint8_t sender_id __attribute__((unused)), const float *distance)
+{
+  if (distance > 0) {
+    estimator_agl = *distance;
```

Comment:
In this particular case, the data is read only and the copy is atomic, so it will be thread safe.
But testing the pointer is not correct ;)
```
if (*distance > 0) {
```